Add details to lookup rules for for extensions (#7179)
dotnet/csharplang: proposals/extensions.md

  over extension type members available in outer namespace declarations,
- that extension methods available in inner namespace declarations take precedence
  over extension methods available in outer namespace declarations, 
- that extension type members declared directly in a namespace take precedence
  over extension type members imported into that same namespace with a using namespace directive,
- and that extension methods declared directly in a namespace take precedence
  over extension methods imported into that same namespace with a using namespace directive.

TODO clarify behavior for extension on `object` or `dynamic` used as `dynamic.M()`?

### Indexer access

TL;DR: If no candidate is applicable, then we attempt extension indexer access instead.

We modify the [element access rules](https://github.com/dotnet/csharpstandard/blob/draft-v7/standard/expressions.md#128111-general) as follows:

/[...]
An *element_access* is dynamically bound if \[...]
If the *primary_no_array_creation_expression* of an *element_access* is a value of an *array_type*, the *element_access* is an array access. 

> Otherwise, the *primary_no_array_creation_expression* shall be a variable or value of a class, struct, or interface type 

Reviewer: The if/otherwise/otherwise wording doesn't make sense, the first two cases sound like they together already cover everything, should be probably if/else/otherwise.

```suggestion
Else if the *primary_no_array_creation_expression* is a variable or value of a class, struct, or interface type 
```

Author: This is existing spec language. I agree it's probably not formulated the best...
The changes I'm making are highlighted with `**`. I just provided this as context to show where the insertion goes.

Reviewer: I know, I think it was formulated fine before you added the second "Otherwise" clause - then I would change the existing "Otherwise" to something else.

that has one or more indexer members, in which case the *element_access* is an indexer access.
\***Otherwise, the *primary_no_array_creation_expression* shall be a variable or value of a class, struct, or interface type 
that has no indexer members, in which case the *element_access* is an extension indexer access.**

We modify the [indexer access rules](https://github.com/dotnet/csharpstandard/blob/draft-v7/standard/expressions.md#128113-indexer-access) as follows:

/[...]

The binding-time processing of an indexer access of the form `P[A]`, where `P` is a *primary_no_array_creation_expression* of a class, struct, or interface type `T`, and `A` is an *argument_list*, consists of the following steps:

- The set of indexers provided by `T` is constructed. \[...]
- The set is reduced to those indexers that are applicable and not hidden by other indexers. \[...]
- \***If the resulting set of candidate indexers is empty, then further processing 
  along the following steps are abandoned, and instead an attempt is made 
  to process the indexer access as an extension indexer access. If this fails, 
  then no applicable indexers exist, and a binding-time error occurs.**

> - If the resulting set of candidate indexers is empty, then no applicable indexers exist, and a binding-time error occurs.

Reviewer: Seems like this step is covered in the previous one; should it be removed?

- The best indexer of the set of candidate indexers is identified using the overload resolution rules. If a single best indexer cannot be identified, the indexer access is ambiguous, and a binding-time error occurs.
- /[...]

/[...]

#### Extension indexer access

TODO3 Confirm what expressions can be the receiver. How about `base`?

In an element access of one of the forms

```csharp
«expr» [ ]
«expr» [ «args» ]
```

if the normal processing of the element access finds no applicable indexers, an attempt is made to process the construct as an extension indexer access. If «expr» or any of the «args» has compile-time type `dynamic`, extension methods will not apply.

This succeeds if, given that «expr» has type `Type`, we find
a substituted compatible implicit extension type `X` for `Type`
so that the corresponding element access can take place:
```csharp
((X)expr) . «identifier» [ ]
((X)expr) . «identifier» [ «args» ]
```

The search proceeds as follows:

> - Starting with the closest enclosing type declaration, continuing with each type declaration,

Reviewer: ```suggestion
- Starting with the closest enclosing type declaration, continuing with each enclosing type declaration,
```

- If `T` is an *interface_type*, the base types of `T` are the base interfaces of `T` and the class type `object`.
- If `T` is an *array_type*, the base types of `T` are the class types `System.Array` and `object`.
- If `T` is a *delegate_type*, the base types of `T` are the class types `System.Delegate` and `object`.

> - \***If `T` is an *extension_type*, the base types of `T` are the base extensions of `T` and the extended type of `T`.**

Reviewer: What about the extended types of the base extensions - they can be different, right?

Also PR https://github.com/dotnet/roslyn/pull/71779 has "and its base types." in the spec quote which isn't here.
